### PR TITLE
Add comment to XContentBuilder.toString()

### PR DESCRIPTION
**Pull request by chishui**
### Description
Add comments to XContentBuilder.toString() to call out that calling this method could close the builder.

### Related Issues
Resolves https://github.com/opensearch-project/OpenSearch/issues/13731

### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [x] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [x] Commits are signed per the DCO using --signoff
- [x] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by chishui** [comment]
Changelog is not needed.

File: libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java [hunk 1]
```diff
@@ -157,6 +157,9 @@ public static XContentBuilder builder(XContent xContent, Set<String> includes, S
 
     /**
      * Returns a string representation of the builder (only applicable for text based xcontent).
+     * Note: explicitly or implicitly (from debugger) calling toString() could cause XContentBuilder
+     * to close which is a side effect done by @see BytesReference#bytes().
+     * Trying to write more contents after toString() will cause NPE. Use it with cautious.
```
Comment (dblock) [inline libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java:162]: Typo: with caution

**Review by dblock — CHANGES_REQUESTED** [review]

File: libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java (hunk 1 above)
Comment (chishui) [reply libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java]: Fixed

File: libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java [hunk 2]
```diff
@@ -157,6 +157,9 @@ public static XContentBuilder builder(XContent xContent, Set<String> includes, S
 
     /**
      * Returns a string representation of the builder (only applicable for text based xcontent).
+     * Note: explicitly or implicitly (from debugger) calling toString() could cause XContentBuilder
```
Comment (reta) [inline libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java:160]: Thanks for the comment, @chishui , pushing it further: should we replace `toString` impl to not dump the content, the method is clearly dangerous to use.

Comment (chishui) [reply libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java:160]: Ideally, we should, but if you search "builder.toString()", you'll find many implementation which simply `return builder.toString()` without closing it. There could be similar usage in plugins as well. Seems to be pretty big effort and could bring potential risk.

Comment (reta) [reply libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java:160]: Fair point, we may be indeed constrained by what we could do in 2.x without breaking things (updating javadocs is OK), but 3.x is certainly the place where  we may try to get rid of such dangerous API

**Review by andrross — APPROVED** [review]

File: libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java (hunk 2 above)
Comment (andrross) [reply libs/core/src/main/java/org/opensearch/core/xcontent/XContentBuilder.java:160]: I think this documentation update is the right first step. I would also support changing the behavior in 3.x, though I'm not super concerned about this behavior. To me, asking to see the contents of a builder, even in toString(), is somewhat akin to "building" it and it's not crazy to expect the builder to no longer be mutable after that operation. But I get that it is unintuitive.

**Comment by andrross** [comment]
> ❌ Gradle check result for [5cc0a04](https://github.com/opensearch-project/OpenSearch/commit/5cc0a04037fbafaa33611a59bd8bd258d5fc2655): [FAILURE](https://build.ci.opensearch.org/job/gradle-check/39264/)

#10006
#13473

**Comment by chishui** [comment]
@dblock the change your requested has been addressed, could you take a look?

**Review by dblock — APPROVED** [review]

**Pull request merged**